warehouse/accounts/services.py

```diff
@@ -618,6 +623,78 @@ def get_password_timestamp(self, user_id):
         user = self.get_user(user_id)
         return user.password_date.timestamp() if user.password_date is not None else 0
 
+    def check_device_valid(self, user_id: str, device_id_secret_b64: str) -> bool:
+        """
+        Checks whether the given device id and secret are valid for the given user.
+        """
+        self._metrics.increment("warehouse.authentication.check_device_valid.start")
+
+        if not device_id_secret_b64:
+            return False
+        try:
+            device_id_secret = DeviceIdSecret.from_base64(device_id_secret_b64)
+        except (KeyError, ValueError):
```

Comment:
I can't off the top of my head remember what all the juicy Exceptions we could expect from `json.loads(base64.b64decode(b64).decode("utf-8"))` are. I think we should probably add a parameterized test with a bunch of garbage values to exercise this. Off the top of my head I think a string that has extended unicode would raise an exception we wouldn't catch here and lead to a 500.

See https://github.com/pypi/warehouse/blob/46a15cc902531e156afb29be5ad6463565eaec84/tests/unit/macaroons/test_services.py#L132-L145 for a concise example of a similar concept.

Reply:
Good idea!